File: Mathematics/Inclusion-Exclusion Principle/atcoder/B31/B31.php
<!-- 以下は、**PHP 8.2.8** を用いた実装です。
`1 以上 N 以下の整数のうち、3, 5, 7 のいずれかで割り切れる数の個数` を高速に求めるため、**包除原理**を用いた定数時間解法となっています。

---

## ✅ 解法：包除原理（Inclusion-Exclusion）

集合 A: 3で割り切れる数
集合 B: 5で割り切れる数
集合 C: 7で割り切れる数

```text
|A ∪ B ∪ C| = |A| + |B| + |C| - |A ∩ B| - |A ∩ C| - |B ∩ C| + |A ∩ B ∩ C|
```

---

## ✅ PHP 実装（型明示 + コメント付き）

```php -->
<?php

/**
 * N 以下の整数のうち、3,5,7 のいずれかで割り切れる個数を返す
 *
 * @param int|string $N 1以上10^12以下の整数（10^12 を超える可能性を考慮し string も許容）
 * @return string 条件を満たす個数（大きな値のため string で返す）
 */
function countMultiplesOf3or5or7(int|string $N): string {
    // 任意精度整数に変換（bcmath 使用）
    $N = (string)$N;

    // 小さな関数：floor(N / x)
    $div = function(string $x) use ($N): string {
        return bcdiv($N, $x, 0); // 0桁 → floor動作
    };

    // Inclusion-Exclusion
    $result = bcadd(bcadd($div("3"), $div("5")), $div("7")); // +A+B+C
    $result = bcsub($result, $div(bcmul("3", "5")));         // -AB
    $result = bcsub($result, $div(bcmul("3", "7")));         // -AC
    $result = bcsub($result, $div(bcmul("5", "7")));         // -BC
    $result = bcadd($result, $div(bcmul("105", "1")));       // +ABC

    return $result;
}

// 標準入力から N を取得
$N = trim(fgets(STDIN));
echo countMultiplesOf3or5or7($N) . PHP_EOL;

// ## ✅ 実行方法

// ```bash
// php main.php < input.txt
// ```

// ---

// ## ✅ 入力例と出力例

// #### 入力:

// ```
// 10
// ```

// #### 出力:

// ```
// 6
// ```

// ---

// ## ✅ 実行時間・メモリ消費量

// | 項目         | 内容                               |
// | ---------- | -------------------------------- |
// | **計算量**    | O(1)（bcmath の除算・加算のみ）            |
// | **実行時間**   | 約 0.1 ～ 1.0 ms（bcmath使用のため少し遅め）  |
// | **メモリ使用量** | 非常に小さい（約 1KB 未満）                 |
// | **精度**     | `bcmath` により任意精度対応 (`N ≤ 10^12`) |

// ---

// ## ✅ 注意点

// * `int` 型は 64bit 環境であっても `10^12` に対し桁数制限があるため、**`bcmath` 拡張**を使って任意精度計算を行っています。
// * 返却型は **`string`** です（bcmathの仕様上）。

// ---

// ご希望があれば、`gmp` 拡張での高速版や、逐次的なループによるナイーブ解法もご提示できます。

File: Mathematics/Inclusion-Exclusion Principle/atcoder/B31/B31.go
// 以下は、**Go (1.20.6)** を用いた `1 以上 N 以下の整数のうち、3, 5, 7 のいずれかで割り切れる数の個数` を求める実装です。

// ---

// ## ✅ 解法：包除原理（Inclusion-Exclusion）

// 次の式により、重複を避けた個数を高速に計算できます：

// ```text
// |A ∪ B ∪ C| = |A| + |B| + |C| - |A ∩ B| - |A ∩ C| - |B ∩ C| + |A ∩ B ∩ C|
// ```

// ---

// ## ✅ Go 実装（型・処理時間・メモリを意識）

// ```go
package main

import (
	"bufio"
	"fmt"
	"math/big"
	"os"
	"strings"
)

/*
	countMultiplesOf3or5or7 は、N 以下の整数のうち、
	3, 5, 7 のいずれかで割り切れる数の個数を返します。

	@param N *big.Int : 任意精度の整数 N (1 ≤ N ≤ 10^12)
	@return *big.Int : 条件を満たす個数（任意精度整数）
*/
func countMultiplesOf3or5or7(N *big.Int) *big.Int {
	// 補助関数: floor(N / x)
	div := func(x int64) *big.Int {
		return new(big.Int).Div(N, big.NewInt(x))
	}

	// 各項を計算（包除原理）
	a := div(3)
	b := div(5)
	c := div(7)

	ab := div(3 * 5)
	ac := div(3 * 7)
	bc := div(5 * 7)

	abc := div(3 * 5 * 7)

	// A + B + C - AB - AC - BC + ABC
	result := new(big.Int).Add(a, b)
	result.Add(result, c)
	result.Sub(result, ab)
	result.Sub(result, ac)
	result.Sub(result, bc)
	result.Add(result, abc)

	return result
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	line = strings.TrimSpace(line)

	N := new(big.Int)
	N.SetString(line, 10) // 入力を任意精度整数に変換

	result := countMultiplesOf3or5or7(N)
	fmt.Println(result.String())
}

// ## ✅ 入出力例

// ```txt
// 入力:
// 210

// 出力:
// 114
// ```

// ---

// ## ✅ 実行方法

// ファイル名を `main.go` として保存し、以下のコマンドで実行できます。

// ```bash
// go run main.go < input.txt
// ```

// ---

// ## ✅ 処理時間・メモリ使用量（想定）

// | 指標     | 内容                             |
// | ------ | ------------------------------ |
// | 時間計算量  | O(1)（定数回の除算 + 加減算のみ）           |
// | 実行時間目安 | 約 0.05ms ～ 0.2ms（N の大きさに無関係）   |
// | メモリ消費  | 数個の `*big.Int` のみ（≒ 数 KB 未満）   |
// | 精度・安全性 | `math/big` により最大 `10^1000` でも可 |

// ---

// ## ✅ コメント

// * Go の `int64` 型は `10^12` まで対応できますが、整数の積（105）との除算時に桁あふれを避けるため、安全策として `math/big.Int` による任意精度処理を採用しています。
// * もし `N <= 10^18` を保証できる環境で高速化したい場合は、`int64` ベースの処理に書き換えも可能です。

// ---

// ご希望に応じて、**`int64` ベース版・ループ版・テスト付きのGoコード**も提供可能です。
